The serialization needs some work for this approach. Here is an incomplete test:

```python
def test_mapped_decorator_serde():
    from airflow.models.xcom_arg import XComArg
    from airflow.decorators import task

    with DAG("test-dag", start_date=datetime(2020, 1, 1)) as dag:
        task1 = BaseOperator(task_id="op1")
        xcomarg = XComArg(task1, "test_key")

        @task(retry_delay=30)
        def x(arg1, arg2):
            ...

    real_op = x.partial(arg1=1).map(arg2=xcomarg).operator

    serialized = SerializedBaseOperator._serialize(real_op)

    assert serialized == {
        '_is_dummy': False,
        '_is_mapped': True,
        '_task_module': '???',
        '_task_type': '???',
        'downstream_task_ids': [],
        'mapped_kwargs': {
            'arg1': [
                1,
                2,
                {"__type": "dict", "__var": {'a': 'b'}},
            ],
            'arg2': {'__type': 'xcomref', '__var': {'task_id': 'op1', 'key': 'test_key'}},
        },
        'task_id': 'x',
        # We don't want to include the python source code in the serialized representation
        # TODO? Where does `retry_delay` go? We need to separate 
    }

    op = SerializedBaseOperator.deserialize_operator(serialized)
    assert isinstance(op, MappedOperator)
    assert op.deps is MappedOperator.DEFAULT_DEPS

    # TODO: add some more asserts here
```

Trying to call `real_op.unmap()` also throws a key error:


```
tests/serialization/test_dag_serialization.py:1670: in test_mapped_decorator_serde
    real_op.unmap()
airflow/models/baseoperator.py:1882: in unmap
    dag._remove_task(self.task_id)
airflow/models/dag.py:2167: in _remove_task
    task = self.task_dict.pop(task_id)
E   KeyError: 'x'
```

--- a/airflow/decorators/base.py
+++ b/airflow/decorators/base.py
@@ -280,30 +280,70 @@ class _TaskDecorator(Generic[Function, OperatorSubclass]):
             names = ", ".join(repr(n) for n in unknown_args)
             raise TypeError(f'{funcname} got unexpected keyword arguments {names}')
 
-    def map(
-        self, *, dag: Optional["DAG"] = None, task_group: Optional["TaskGroup"] = None, **kwargs
-    ) -> XComArg:
+    def map(self, *args, **kwargs) -> XComArg:
         self._validate_arg_names("map", kwargs)
-        dag = dag or DagContext.get_current_dag()
-        task_group = task_group or TaskGroupContext.get_current_task_group(dag)
-        task_id = get_unique_task_id(self.kwargs['task_id'], dag, task_group)
 
-        operator = MappedOperator.from_decorator(
-            decorator=self,
+        partial_kwargs: Dict[str, Any] = {
+            **self.kwargs,
+            "python_callable": self.function,
+            "multiple_outputs": self.multiple_outputs,
+        }
+
+        dag = partial_kwargs.pop("dag", DagContext.get_current_dag())
+        task_group = partial_kwargs.pop("task_group", TaskGroupContext.get_current_task_group(dag))
+        task_id = get_unique_task_id(partial_kwargs.pop("task_id"), dag, task_group)
+
+        operator = MappedOperator(
+            operator_class=self.operator_class,
+            partial_kwargs=partial_kwargs,
+            mapped_kwargs={},
+            task_id=task_id,
             dag=dag,
             task_group=task_group,
-            task_id=task_id,
-            mapped_kwargs=kwargs,
+            deps=MappedOperator._deps(self.operator_class.deps),
         )
+        operator.mapped_kwargs["op_args"] = list(args)
+        operator.mapped_kwargs["op_kwargs"] = kwargs
+
+        for arg in itertools.chain(args, kwargs.values()):
+            XComArg.apply_upstream_relationship(operator, arg)
         return XComArg(operator=operator)
 
-    def partial(
-        self, *, dag: Optional["DAG"] = None, task_group: Optional["TaskGroup"] = None, **kwargs
-    ) -> "_TaskDecorator[Function, OperatorSubclass]":
-        self._validate_arg_names("partial", kwargs, {'task_id'})
-        partial_kwargs = self.kwargs.copy()
-        partial_kwargs.update(kwargs)
-        return attr.evolve(self, kwargs=partial_kwargs)
+    def partial(self, *args, **kwargs) -> "_TaskDecorator[Function, OperatorSubclass]":
+        self._validate_arg_names("partial", kwargs)
+
+        op_args = self.kwargs.get("op_args", [])
+        op_args.extend(args)
+
+        op_kwargs = self.kwargs.get("op_kwargs", {})
+        duplicated_keys = set(op_kwargs).intersection(kwargs)
+        if len(duplicated_keys) == 1:
+            raise TypeError(f"duplicated partial argument: {duplicated_keys.pop()}")
+        elif duplicated_keys:
+            duplicated_keys_display = ", ".join(sorted(duplicated_keys))
+            raise TypeError(f"duplicated partial arguments: {duplicated_keys_display}")
+        op_kwargs.update(kwargs)
+
+        return attr.evolve(self, kwargs={**self.kwargs, "op_args": op_args, "op_kwargs": op_kwargs})
+
+
+class Task(Generic[Function]):
+    """Declaration of a @task-decorated callable for type-checking.
+
+    An instance of this type inherits the call signature of the decorated
+    function wrapped in it (not *exactly* since it actually returns an XComArg,
+    but there's no way to express that right now), and provides two additional
+    methods for task-mapping.
+
+    This type is implemented by ``_TaskDecorator`` at runtime.
+    """
+
+    __call__: Function
+
+    function: Function
+
+    map: Callable[..., XComArg]
+    partial: Callable[..., "Task[Function]"]
 
 
 class Task(Generic[Function]):

--- a/airflow/models/baseoperator.py
+++ b/airflow/models/baseoperator.py
@@ -82,7 +82,6 @@ from airflow.utils.trigger_rule import TriggerRule
 from airflow.utils.weight_rule import WeightRule
 
 if TYPE_CHECKING:
-    from airflow.decorators.base import _TaskDecorator
     from airflow.models.dag import DAG
     from airflow.utils.task_group import TaskGroup
 
@@ -243,7 +242,7 @@ class BaseOperatorMeta(abc.ABCMeta):
         return new_cls
 
     # The class level partial function. This is what handles the actual mapping
-    def partial(cls, *, task_id: str, dag: Optional["DAG"] = None, **kwargs):
+    def partial(cls, *, task_id: str, dag: Optional["DAG"] = None, **kwargs) -> "MappedOperator":
         operator_class = cast("Type[BaseOperator]", cls)
         # Validate that the args we passed are known -- at call/DAG parse time, not run time!
         _validate_kwarg_names_for_mapping(operator_class, "partial", kwargs)
@@ -1632,7 +1631,7 @@ class MappedOperator(Operator, LoggingMixin, DAGNode):
             dag._remove_task(operator.task_id)
 
         operator_init_kwargs: dict = operator._BaseOperator__init_kwargs  # type: ignore
-        return MappedOperator(
+        return cls(
             operator_class=type(operator),
             task_id=operator.task_id,
             task_group=task_group,
@@ -1647,37 +1646,6 @@ class MappedOperator(Operator, LoggingMixin, DAGNode):
             max_active_tis_per_dag=operator.max_active_tis_per_dag,
             deps=cls._deps(operator.deps),
         )
-
-    @classmethod
-    def from_decorator(
-        cls,
-        *,
-        decorator: "_TaskDecorator",
-        dag: Optional["DAG"],
-        task_group: Optional["TaskGroup"],
-        task_id: str,
-        mapped_kwargs: Dict[str, Any],
-    ) -> "MappedOperator":
-        """Create a mapped operator from a task decorator.
-
-        Different from ``from_operator``, this DOES NOT validate ``mapped_kwargs``.
-        The task decorator calling this should be responsible for validation.
-        """
-        from airflow.models.xcom_arg import XComArg
-
-        operator = MappedOperator(
-            operator_class=decorator.operator_class,
-            partial_kwargs=decorator.kwargs,
-            mapped_kwargs={},
-            task_id=task_id,
-            dag=dag,
-            task_group=task_group,
-            deps=cls._deps(decorator.operator_class.deps),
-        )
-        operator.mapped_kwargs.update(mapped_kwargs)
-        for arg in mapped_kwargs.values():
-            XComArg.apply_upstream_relationship(operator, arg)
-        return operator
 
     @classmethod
     def _deps(cls, deps: Iterable[BaseTIDep]):
@@ -1902,22 +1870,17 @@ class MappedOperator(Operator, LoggingMixin, DAGNode):
 
         return ret
 
+    def create_unmapped_operator(self, dag: "DAG", kwargs: Dict[str, Any]) -> BaseOperator:
+        assert not isinstance(self.operator_class, str)
+        return self.operator_class(dag=dag, task_id=self.task_id, **kwargs)
+
     def unmap(self) -> BaseOperator:
         """Get the "normal" Operator after applying the current mapping"""
-        assert not isinstance(self.operator_class, str)
-
         dag = self.get_dag()
         if not dag:
-            raise RuntimeError("Cannot unmapp a task unless it has a dag")
-
-        args = {
-            **self.partial_kwargs,
-            **self.mapped_kwargs,
-        }
+            raise RuntimeError("Cannot unmap a task unless it has a DAG")
         dag._remove_task(self.task_id)
-        task = self.operator_class(task_id=self.task_id, dag=self.dag, **args)
-
-        return task
+        return self.create_unmapped_operator(dag, {**self.partial_kwargs, **self.mapped_kwargs})
 
 
 # TODO: Deprecate for Airflow 3.0

--- a/airflow/models/taskinstance.py
+++ b/airflow/models/taskinstance.py
@@ -1713,7 +1713,7 @@ class TaskInstance(Base, LoggingMixin):
         test_mode: Optional[bool] = None,
         force_fail: bool = False,
         error_file: Optional[str] = None,
-        session=NEW_SESSION,
+        session: Session = NEW_SESSION,
     ) -> None:
         """Handle Failure for the TaskInstance"""
         if test_mode is None:

--- a/tests/dags/test_mapped_taskflow.py
+++ b/tests/dags/test_mapped_taskflow.py
@@ -1,0 +1,31 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow import DAG
+from airflow.utils.dates import days_ago
+
+with DAG(dag_id='test_mapped_taskflow', start_date=days_ago(2)) as dag:
+
+    @dag.task
+    def make_list():
+        return [1, 2, {'a': 'b'}]
+
+    @dag.task
+    def consumer(value):
+        print(repr(value))
+
+    consumer.map(value=make_list())

--- a/tests/decorators/test_python.py
+++ b/tests/decorators/test_python.py
@@ -490,7 +490,7 @@ def test_mapped_decorator() -> None:
     assert isinstance(doubled_0, XComArg)
     assert isinstance(doubled_0.operator, MappedOperator)
     assert doubled_0.operator.task_id == "double"
-    assert doubled_0.operator.mapped_kwargs == {"number": literal}
+    assert doubled_0.operator.mapped_kwargs == {"op_args": [], "op_kwargs": {"number": literal}}
 
     assert doubled_1.operator.task_id == "double__1"
 
@@ -514,25 +514,37 @@ def test_partial_mapped_decorator() -> None:
     def product(number: int, multiple: int):
         return number * multiple
 
+    literal = [1, 2, 3]
+
     with DAG('test_dag', start_date=DEFAULT_DATE) as dag:
-        literal = [1, 2, 3]
-        quadrupled = product.partial(task_id='times_4', multiple=3).map(number=literal)
+        quadrupled = product.partial(multiple=3).map(number=literal)
         doubled = product.partial(multiple=2).map(number=literal)
         trippled = product.partial(multiple=3).map(number=literal)
 
-        product.partial(multiple=2)
+        product.partial(multiple=2)  # No operator is actually created.
+
+    assert dag.task_dict == {
+        "product": quadrupled.operator,
+        "product__1": doubled.operator,
+        "product__2": trippled.operator,
+    }
 
     assert isinstance(doubled, XComArg)
     assert isinstance(doubled.operator, MappedOperator)
-    assert doubled.operator.task_id == "product"
-    assert doubled.operator.mapped_kwargs == {"number": literal}
-    assert doubled.operator.partial_kwargs == {"task_id": "product", "multiple": 2}
+    assert doubled.operator.mapped_kwargs == {"op_args": [], "op_kwargs": {"number": literal}}
+    assert doubled.operator.partial_kwargs == {
+        "python_callable": product.function,
+        "multiple_outputs": False,
+        "op_args": [],
+        "op_kwargs": {"multiple": 2},
+    }
 
-    assert trippled.operator.task_id == "product__1"
-    assert trippled.operator.partial_kwargs == {"task_id": "product", "multiple": 3}
-
-    assert quadrupled.operator.task_id == "times_4"
+    assert isinstance(trippled.operator, MappedOperator)  # For type-checking on partial_kwargs.
+    assert trippled.operator.partial_kwargs == {
+        "python_callable": product.function,
+        "multiple_outputs": False,
+        "op_args": [],
+        "op_kwargs": {"multiple": 3},
+    }
 
     assert doubled.operator is not trippled.operator
-
-    assert [quadrupled.operator, doubled.operator, trippled.operator] == dag.tasks

--- a/tests/jobs/test_backfill_job.py
+++ b/tests/jobs/test_backfill_job.py
@@ -1519,13 +1519,14 @@ class TestBackfillJob:
         job.run()
         assert executor.job_id is not None
 
-    def test_mapped_dag(self):
+    @pytest.mark.parametrize("dag_id", ["test_mapped_classic", "test_mapped_taskflow"])
+    def test_mapped_dag(self, dag_id):
         """End-to-end test of a simple mapped dag"""
         # Use SequentialExecutor for more predictable test behaviour
         from airflow.executors.sequential_executor import SequentialExecutor
 
-        self.dagbag.process_file(str(TEST_DAGS_FOLDER / 'test_mapped_classic.py'))
-        dag = self.dagbag.get_dag('test_mapped_classic')
+        self.dagbag.process_file(str(TEST_DAGS_FOLDER / f'{dag_id}.py'))
+        dag = self.dagbag.get_dag(dag_id)
 
         # This needs a real executor to run, so that the `make_list` task can write out the TaskMap
 

--- a/tests/jobs/test_backfill_job.py
+++ b/tests/jobs/test_backfill_job.py
@@ -47,7 +47,13 @@ from airflow.utils.state import State
 from airflow.utils.timeout import timeout
 from airflow.utils.types import DagRunType
 from tests.models import TEST_DAGS_FOLDER
-from tests.test_utils.db import clear_db_dags, clear_db_pools, clear_db_runs, set_default_pool_slots
+from tests.test_utils.db import (
+    clear_db_dags,
+    clear_db_pools,
+    clear_db_runs,
+    clear_db_xcom,
+    set_default_pool_slots,
+)
 from tests.test_utils.mock_executor import MockExecutor
 from tests.test_utils.timetables import cron_timetable
 
@@ -66,6 +72,7 @@ class TestBackfillJob:
     def clean_db():
         clear_db_dags()
         clear_db_runs()
+        clear_db_xcom()
         clear_db_pools()
 
     @pytest.fixture(autouse=True)
@@ -1512,7 +1519,7 @@ class TestBackfillJob:
         job.run()
         assert executor.job_id is not None
 
-    def test_mapped_dag(self, dag_maker):
+    def test_mapped_dag(self):
         """End-to-end test of a simple mapped dag"""
         # Use SequentialExecutor for more predictable test behaviour
         from airflow.executors.sequential_executor import SequentialExecutor

